src/endpoints/users.js

```diff
@@ -22,6 +22,24 @@ const visibleGroups = (user) => ALL_GROUPS
   }));
 
 
+/**
+ * Returns an array of Nextstrain groups that a given *user* is a member of. The
+ * order of groups returned matches the order in the `groups.json` data file.
+ *
+ * @param {Object} user
+ * @returns {Array} Each element is an object with keys `name` -> {str} (group name) and
+ *                  `private` -> {bool} (group is private)
+ */
+const groupMemberships = (user) => ALL_GROUPS
+  .map(g => [g.name, g.source])
+  // Check roles (formatted as '<groupName>/<role>') for group membership.
+  .filter(([name ,]) => [...user.authzRoles].some(role => role.split('/')[0] === name))
+  .map(([name, source]) => ({
+    name,
+    private: !authz.authorized(null, authz.actions.Read, source),
+  }));
```

Comment:
Code outside of `src/authn/` (and `src/groups.js`) generally shouldn't be splitting (or joining) on `/` to form authz roles. That implementation detail should remain encapsulated. There's also a list of the user's groups (by name) already on the user object, so filtering down in this way from `ALL_GROUPS` isn't necessary.

```suggestion
const groupMemberships = (user) => user?.groups?
  .map(name => new Group(name))
  .map(group => ({
    name: group.name,
    private: !group.isPublic,
  }))
  ??
  []
;
```

The nullish coalescing operators used make it so the caller can simply do `groupMemberships(req.user)` even when `req.user === null`.

Longer term, it probably makes sense to convert `req.user.groups` to Group objects in the authn code rather than here. There can also be some additional error/edge-case handling (e.g. group names reported by Cognito that the server doesn't know about).

Reply:
Good suggestion. Switched to use `user.groups` in 3a4dcc79a8978bda401726d43c7caf7b72dc9c5a. I took it a bit further by returning the list of Groups to use `name` and `isPublic` directly instead of through an intermediate object.